dataset/dataset.go

```diff
@@ -467,15 +477,68 @@ func (c *Client) GetVersions(ctx context.Context, userAuthToken, serviceAuthToke
 		return
 	}
 
-	versions := struct {
-		Items []Version `json:"items"`
-	}{}
+	if err = json.Unmarshal(b, &m); err != nil {
+		return
+	}
 
-	err = json.Unmarshal(b, &versions)
-	m = versions.Items
 	return
 }
 
+// GetVersionsInBatches retrieves a list of datasets in concurrent batches and accumulates the results
+func (c *Client) GetVersionsInBatches(ctx context.Context, userAuthToken, serviceAuthToken, downloadServiceAuthToken, collectionID, datasetID, edition string, batchSize, maxWorkers int) (versions VersionsList, err error) {
+
+	// Function to aggregate items.
+	// For the first received batch, as we have the total count information, will initialise the final structure of items with a fixed size equal to TotalCount.
+	// This serves two purposes:
+	//   - We can guarantee, even with concurrent calls, that values are returned in the same order that the API defines, by offsetting the index.
+	//   - We do a single memory allocation for the final array, making the code more memory efficient.
+	var processBatch VersionsBatchProcessor = func(b VersionsList) (abort bool, err error) {
+		if len(versions.Items) == 0 { // first batch response being handled
+			versions.TotalCount = b.TotalCount
+			versions.Items = make([]Version, b.TotalCount)
+			versions.Count = b.TotalCount
+		}
+		for i := 0; i < len(b.Items); i++ {
+			if len(versions.Items) < i+b.Offset {
```

Comment:
Sorry I should have spotted this during the chat but if you update this to:

```
if len(versions.Items) < len(b.Items) + b.Offset {
    return false, fmt.Errorf(....)
}
```
You can put it before the loop and it will be more efficient than checking in each iteration.